crates/core/src/nucleus/validation/build_from_dht.rs

```diff
@@ -0,0 +1,251 @@
+use crate::{
+    context::Context, entry::CanPublish, network::entry_with_header::EntryWithHeader,
+    workflows::get_entry_result::get_entry_with_meta_workflow,
+};
+use holochain_core_types::{
+    chain_header::ChainHeader,
+    entry::{Entry, EntryWithMeta, EntryWithMetaAndHeader},
+    error::HolochainError,
+    time::Timeout,
+    validation::{ValidationPackage, ValidationPackageDefinition},
+};
+
+use std::sync::Arc;
+
+const GET_TIMEOUT_MS: usize = 1000;
+
+async fn all_chain_headers_before_header_dht(
+    context: Arc<Context>,
+    header: &ChainHeader,
+) -> Result<Vec<ChainHeader>, HolochainError> {
+    let mut current_header = header.clone();
+    let mut headers = Vec::new();
+
+    while let Some(next_header_addr) = current_header.link() {
+        log_debug!(context, "About to try and get header: {}", next_header_addr);
+
+        let timeout = Timeout::new(GET_TIMEOUT_MS);
+        // when trying to run the line below there is a panic because the state is dropped
+        // this does not happen when running the tests below. Only when triggered as part of a
+        // validate entry workflow
+        let get_entry_result =
+            get_entry_with_meta_workflow(&context, &next_header_addr, &timeout).await;
+
+        log_debug!(
+            context,
+            "get_entry_workflow returned: {:?}",
+            get_entry_result
+        );
+
+        if let Ok(Some(EntryWithMetaAndHeader {
+            entry_with_meta:
+                EntryWithMeta {
+                    entry: Entry::ChainHeader(chain_header),
+                    ..
+                },
+            ..
+        })) = get_entry_result
+        {
+            headers.push(chain_header.clone());
+            current_header = chain_header;
+        } else {
+            log_debug!(context, "When building validation package from DHT, Could not retrieve a header entry at address: {:?}. Got {:?}", next_header_addr, get_entry_result);
+            return Err(HolochainError::ErrorGeneric(
+                format!("When building validation package from DHT, Could not retrieve a header entry at address: {:?}", next_header_addr))
```

Comment:
Can we add the Error reason in here too, because this is what may be reported back out of zome calls and pouring through the DEBUG logs would be difficult.